airbyte-ci/connectors/pipelines/pipelines/helpers/run_steps.py

```diff
@@ -0,0 +1,233 @@
+        else:
+            raise Exception(f"Unexpected step type: {type(runnable_step)}")
+
+    return skipped_results
+
+def _step_dependencies_succeeded(depends_on: List[str], results: RESULTS_DICT) -> bool:
+    """
+    Check if all dependencies of a step have succeeded.
+    """
+    main_logger.info(f"Checking if dependencies {depends_on} have succeeded")
+    return all(results.get(step_id) and results.get(step_id).status is StepStatus.SUCCESS for step_id in depends_on)
+
+def _filter_skipped_steps(steps_to_evaluate: STEP_TREE, skip_steps: List[str], results: RESULTS_DICT) -> Tuple[STEP_TREE, RESULTS_DICT]:
+    """
+    Filter out steps that should be skipped.
+
+    Either because they are in the skip list or because one of their dependencies failed.
+    """
+    steps_to_run = []
+    for step_to_eval in steps_to_evaluate:
+
+        # ignore nested steps
+        if isinstance(step_to_eval, list):
+            steps_to_run.append(step_to_eval)
+            continue
+
+        # skip step if its id is in the skip list
+        if step_to_eval.id in skip_steps:
+            main_logger.info(f"Skipping step {step_to_eval.id}")
+            results[step_to_eval.id] = step_to_eval.step.skip("Skipped by user")
+
+        # skip step if a dependency failed
+        elif not _step_dependencies_succeeded(step_to_eval.depends_on, results):
+            main_logger.info(f"Skipping step {step_to_eval.id} because one of the dependencies have not been met: {step_to_eval.depends_on}")
+            results[step_to_eval.id] = step_to_eval.step.skip("Skipped because a dependency was not met")
+
+        else:
+            steps_to_run.append(step_to_eval)
+
+    return steps_to_run, results
+
+def _get_next_step_group(steps: STEP_TREE) -> Tuple[STEP_TREE, STEP_TREE]:
+    """
+    Get the next group of steps to run concurrently.
+    """
+    if not steps:
+        return [], []
+
+    if isinstance(steps[0], list):
+        return steps[0], steps[1:]
+    else:
+        # Termination case: if the next step is not a list that means we have reached the max depth
+        return steps, []
+
+def _log_step_tree(step_tree: STEP_TREE, options: RunStepOptions, depth: int = 0):
+    """
+    Log the step tree to the console.
+
+    e.g.
+    Step tree
+    - step1
+    - step2
+        - step3
+        - step4 (skip)
+            - step5
+    - step6
+    """
+    indent = "    "
+    for steps in step_tree:
+        if isinstance(steps, list):
+            main_logger.info(f"{indent * depth}- {steps[0].id} (parallel)")
+            _log_step_tree(steps, options, depth + 1)
+        else:
+            if steps.id in options.skip_steps:
+                main_logger.info(f"{indent * depth}- {steps.id} (skip)")
+            else:
+                main_logger.info(f"{indent * depth}- {steps.id}")
+
+
+async def run_steps(
+    runnables: STEP_TREE,
+    results: RESULTS_DICT = {},
+    options: RunStepOptions = RunStepOptions(),
```

Comment:
Shall we add a `concurrency` attribute to `RunStepOptions` which would lead to the definition of a semaphore in this function? We currently control concurrency at the connector level but if `steps_to_run` is huge we might like to cap the concurrency.

Reply:
Yes, you'd instantiate a Semaphore with the maximum capacity in the `RunStepOptions`
and wrap the task group on which you'd like to control the concurrency with  `with options.semaphore`